### PR TITLE
feat: composite messages

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

list of changes
- #1952 
- #1977 
- #1980 
- #1996 
- #2002 

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [631](https://10.10.124.134/job/AR-build-pipeline/631/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.wire.kalium.logic.data.id.MessageId
+import com.wire.kalium.logic.data.id.QualifiedID
+import com.wire.kalium.logic.data.id.QualifiedIdMapper
+import com.wire.kalium.logic.feature.message.composite.SendButtonActionMessageUseCase
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class CompositeMessagesViewModel @Inject constructor(
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt]: Instead of having a ViewModel for ALL the buttons, you could create one PER button, this way we do not need to work with a map, and passing it alll the way through the composition to the destined button, this is in my opinion something UI should care about, but because we did not have the tool before, to invoke usecases inside classes that were created using remember, we had not choice but to pass this responsibility to the ViewModel on the top of the three. 

But with the Jakubz introduced library we can scope it to the Composable itself.

So in short, this viewmodel,. would hold only the PendingState as Boolean for example. as it would be scoped to that one button.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 2]
```diff
@@ -571,7 +580,9 @@ fun MessageList(
     onSelfDeletingMessageRead: (UIMessage) -> Unit,
     conversationDetailsData: ConversationDetailsData,
     onFailedMessageRetryClicked: (String) -> Unit,
-    onFailedMessageCancelClicked: (String) -> Unit
+    onFailedMessageCancelClicked: (String) -> Unit,
+    onMessageButtonClicked: (messageId: String, buttonId: String) -> Unit,
+    pendingButtonsMap: Map<String, String>
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]: this could be then removed

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 3]
```diff
@@ -460,6 +465,8 @@ private fun ConversationScreenContent(
     onChangeSelfDeletionClicked: () -> Unit,
     onSearchMentionQueryChanged: (String) -> Unit,
     onClearMentionSearchResult: () -> Unit,
+    onMessageButtonClicked: (messageId: String, buttonId: String) -> Unit,
+    pendingButtonsMap: Map<String, String>,
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]: this could be then removed

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt [hunk 4]
```diff
@@ -90,8 +103,78 @@ internal fun MessageBody(
...
+    }
+    buttonList?.also {
+        MessageButtonsContent(
+            messageId = messageId,
+            buttonList = it,
+            onClick = onButtonClick,
+            pendingButton = pendingButton
+        )
+    }
+}
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt:115]: I said it before, but maybe you forgot about it :P I Think instead of passing nullable stuff that sometimes are used and sometimes not, it makes more sense to extract this 
```

 buttonList?.also {
        MessageButtonsContent(
            messageId = messageId,
            buttonList = it,
            onClick = onButtonClick,
            pendingButton = pendingButton
        )
    }
```

to a seperate Composable

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt [hunk 5]
```diff
@@ -475,12 +481,39 @@ private fun MessageContent(
...
+                    isAvailable = !message.isPending && message.isAvailable,
+                    onLongClick = onLongClick,
+                    onOpenProfile = onOpenProfile,
+                    buttonList = messageContent.buttonList,
+                    onButtonClick = onMessageButtonClicked,
+                    messageId = message.header.messageId,
+                    pendingButton = pendingButtonsMap[message.header.messageId]
+                )
+            }
+        }
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt]: This whole thing could be a Composable on its own, that would containt this Column plus this 
```

 buttonList?.also {
        MessageButtonsContent(
            messageId = messageId,
            buttonList = it,
            onClick = onButtonClick,
            pendingButton = pendingButton
        )
    }
}
```

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt [hunk 6]
```diff
@@ -90,8 +103,78 @@ internal fun MessageBody(
...
+
+@SuppressLint("RememberReturnType")
+@Composable
+fun MessageButtonItem(
+    messageId: String,
+    button: MessageButton,
+    pendingButtonId: MessageButtonId?,
+    onClick: ((messageId: String, buttonId: String) -> Unit)?
+) {
+    val onCLick = remember(button.isSelected) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt]: here would be a good place for that single ViewModel I think, each MessageButtonItem, will just execute the use case and hold the pending state

**Comment by AndroidBob** [comment]
Build [638](https://10.10.124.134/job/AR-build-pipeline/638/) **failed**.

**Comment by AndroidBob** [comment]
Build [689](https://10.10.124.134/job/AR-build-pipeline/689/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.4.0(7140078)(7140078)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.4.0(7140078)(7140078)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt [hunk 7]
```diff
@@ -247,10 +247,20 @@ fun MessagePreview.uiLastMessageContent(): UILastMessageContent {
                 is WithUser.TeamMemberRemoved -> UILastMessageContent.None // TODO
                 is WithUser.Text -> UILastMessageContent.SenderWithMessage(
                     sender = userUIText,
-                    message = UIText.DynamicString((content as WithUser.Text).messageBody),
+                    message = (content as WithUser.Text).messageBody.let { UIText.DynamicString(it) },
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt:250]: isn't this crash prone ?

File: app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt [hunk 8]
```diff
@@ -247,10 +247,20 @@ fun MessagePreview.uiLastMessageContent(): UILastMessageContent {
...
                     separator = ": "
                 )
 
+                is WithUser.Composite -> {
+                    val text = (content as WithUser.Composite).messageBody?.let { UIText.DynamicString(it) }
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt:255]: isn;t this crash prone when content is not WithUser.Composite and we try to cast it ?

File: app/src/main/kotlin/com/wire/android/mapper/RegularMessageContentMapper.kt [hunk 9]
```diff
@@ -91,6 +92,33 @@ class RegularMessageMapper @Inject constructor(
             message.deliveryStatus
         )
 
+        is MessageContent.Composite -> {
+            val text = content.textContent?.let { textContent ->
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/mapper/RegularMessageContentMapper.kt:96]: can;t we map it inside the mapper ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt [hunk 10]
```diff
@@ -90,8 +103,59 @@ internal fun MessageBody(
...
+                loading = isPending,
+                text = button.text,
+                onClick = onCLick,
+                state = state
+            )
+            if (index != buttonList.lastIndex) {
+                Spacer(modifier = Modifier.padding(top = dimensions().spacing8x))
+            }
+        }
+    }
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt:158]: looking awesome <3 thanks for the discussion and openess to changes

**Review by trOnk12 — APPROVED** [review]
some tiny tiny last suggetions from my side

**Review by ohassine — APPROVED** [review]

File: app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt (hunk 7 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt:250]: no, since this code will only be executed when the content is Text

File: app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt (hunk 8 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/mapper/MessagePreviewContentMapper.kt:255]: it is impossible to crash this because of the `when` this code will run iff the content is Composite

**Comment by AndroidBob** [comment]
Build [718](https://10.10.124.134/job/AR-build-pipeline/718/) **failed**.

**Comment by AndroidBob** [comment]
Build [719](https://10.10.124.134/job/AR-build-pipeline/719/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.4.0(7142007)(7142007)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.4.0(7142007)(7142007)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [736](https://10.10.124.134/job/AR-build-pipeline/736/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.4.0(7149228)(7149228)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.4.0(7149228)(7149228)-staging-compat.apk)

**Review by saleniuk — APPROVED** [review]
